AppServer/google/appengine/api/logservice/logservice_stub.py

```diff
@@ -20,30 +20,127 @@
 import base64
 import capnp # pylint: disable=unused-import
 import logging
+
+import json
 import logging_capnp
 import socket
 import struct
 import time
 
-
 from collections import defaultdict
+
+import os
+
+from datetime import datetime
+
+import multiprocessing
+
+import threading
 from google.appengine.api import apiproxy_stub
 from google.appengine.api.logservice import log_service_pb
+from google.appengine.api.modules import (
+  get_current_module_name, get_current_version_name
+)
 from google.appengine.runtime import apiproxy_errors
-from Queue import Queue, Empty
+from Queue import Queue, Empty, Full
 
 # Add path to import file_io
 from appscale.common import file_io
 
+
 _I_SIZE = struct.calcsize('I')
 
+LEVELS = {
+  0: 'DEBUG',
+  1: 'INFO',
+  2: 'WARNING',
+  3: 'ERROR',
+  4: 'CRITICAL',
+}
+
+
+class RequestsLogger(threading.Thread):
+
+  FILENAME_TEMPLATE = (
+    '/opt/appscale/logserver/requests-{app}-{service}-{version}-{port}.log'
+  )
+  QUEUE_SIZE = 1024 * 16
+
+  def __init__(self):
+    super(RequestsLogger, self).__init__()
+    self.setDaemon(True)
+    self._logs_queue = multiprocessing.Queue(self.QUEUE_SIZE)
+    self._log_file = None
+    self._shutting_down = False
+
+  def _open_log_file(self, request_info):
+    # Init logger lazily when application info is available
+    app_id = request_info['appId']
+    service_id = request_info['serviceName']
+    version_id = request_info['versionName']
+    port = request_info['port']
+    # Prepare filename
+    filename = self.FILENAME_TEMPLATE.format(
+      app=app_id, service=service_id, version=version_id, port=port)
+    # Open log file
+    self._log_file = open(filename, 'a')
+
+  def run(self):
+    request_info = None
+    while True:
+      try:
+        try:
+          if not request_info:
+            # Get new info from the queue if previous has been saved
+            request_info = self._logs_queue.get()
+          if not request_info and self._shutting_down:
+            return
+          if not self._log_file:
+            self._open_log_file(request_info)
+          self._log_file.write(json.dumps(request_info))
+          self._log_file.write('\n')
+          self._log_file.flush()
+          request_info = None
+
+        except (OSError, IOError):
+          # Close file to reopen it again later
+          logging.exception(
+            'Failed to write request_info to log file\n  Request info: {}'
+            .format(request_info or "-"))
+          log_file = self._log_file
+          self._log_file = None
```

Comment:
Would the Python garbage collector take care of closing the descriptor stored at `log_file`? Otherwise, this seems like it could leave behind open file descriptors. For example, if the disk is full, the `.close()` method may raise an exception, and it appears that a new open file descriptor would be created every 5 seconds.